src/ast/nodes/IfStatement.ts

```diff
@@ -1,73 +1,88 @@
 import MagicString from 'magic-string';
 import { RenderOptions } from '../../utils/renderHelpers';
+import { DeoptimizableEntity } from '../DeoptimizableEntity';
 import { ExecutionPathOptions } from '../ExecutionPathOptions';
 import { EMPTY_IMMUTABLE_TRACKER } from '../utils/ImmutableEntityPathTracker';
-import { EMPTY_PATH, UNKNOWN_VALUE } from '../values';
+import { EMPTY_PATH, LiteralValueOrUnknown, UNKNOWN_VALUE } from '../values';
 import * as NodeType from './NodeType';
 import { ExpressionNode, StatementBase, StatementNode } from './shared/Node';
 
-export default class IfStatement extends StatementBase {
+export default class IfStatement extends StatementBase implements DeoptimizableEntity {
 	type: NodeType.tIfStatement;
 	test: ExpressionNode;
 	consequent: StatementNode;
 	alternate: StatementNode | null;
 
-	private hasUnknownTestValue: boolean;
+	private testValue: LiteralValueOrUnknown;
+	private needsTestValue: boolean;
+
+	bind() {
+		super.bind();
+		if (this.needsTestValue) {
+			this.testValue = UNKNOWN_VALUE;
+			this.needsTestValue = false;
+			this.testValue = this.test.getLiteralValueAtPath(EMPTY_PATH, EMPTY_IMMUTABLE_TRACKER, this);
+		}
+	}
+
+	deoptimize() {
+		this.testValue = UNKNOWN_VALUE;
```

Comment:
Is this guaranteed to always be called after `bind()` so that `needsTestValue` is definitely `false`?

Reply:
Yes it is. `deoptimize` can only happen after the if statement has been passed as a deoptimization target to `this.test.getLiteralValueAtPath` in the last line of `bind`.